pkg/ccl/streamingccl/streamingest/stream_ingestion_dist.go

```diff
@@ -318,10 +324,101 @@ func measurePlanChange(before, after *sql.PhysicalPlan) float64 {
 	return float64(diff) / float64(oldCount)
 }
 
+type partitionWithCandidates struct {
+	partition          streamclient.PartitionInfo
+	closestDestIDs     []base.SQLInstanceID
+	sharedPrefixLength int
+}
+
+type candidatesByPriority []partitionWithCandidates
+
+func (a candidatesByPriority) Len() int      { return len(a) }
+func (a candidatesByPriority) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a candidatesByPriority) Less(i, j int) bool {
+	return a[i].sharedPrefixLength > a[j].sharedPrefixLength
+}
+
+type nodeMatcher struct {
+	destMatchCount map[base.SQLInstanceID]int
+	destNodesInfo  []sql.InstanceLocality
+}
+
+func makeNodeMatcher(destNodesInfo []sql.InstanceLocality) nodeMatcher {
+	return nodeMatcher{
+		destMatchCount: make(map[base.SQLInstanceID]int, len(destNodesInfo)),
+		destNodesInfo:  destNodesInfo,
+	}
+}
+
+func (nm nodeMatcher) destNodeIDs() []base.SQLInstanceID {
+	allDestNodeIDs := make([]base.SQLInstanceID, 0, len(nm.destNodesInfo))
+	for _, info := range nm.destNodesInfo {
+		allDestNodeIDs = append(allDestNodeIDs, info.GetInstanceID())
+	}
+	return allDestNodeIDs
+}
+
+// findSourceNodePriority finds the closest dest nodes for each source node and
+// returns a list of (source node, dest node match candidates) pairs ordered by
+// matching priority. A source node is earlier (higher priority) in the list if
+// it shares more locality tiers with their destination node match candidates.
```

Comment:
It might be nice to spell out the other parts of the sorting procedure in a block comment on the nodeMatcher struct since it is a little unclear from the code just because the full procedure is spread out over three methods.